### PR TITLE
Read commit and branch information from local git commands

### PR DESCRIPTION
**Pull request by timhaines**
This introduces collecting git commit and branch info from the locally run `git` command, if it hasn't already been specified by ENV vars.

The main benefit of this is that the meta-information no longer needs to be collected from GitHub, so source access is no longer required.

### CODE REVIEW

File: src/environment.js [hunk 1]
```diff
@@ -26,6 +37,66 @@ class Environment {
     return null;
   }
 
+  rawCommitData(commitSha) {
+    // Make sure commitSha is only alphanumeric characters to prevent command injection.
+    if (commitSha.length > 100 || !commitSha.match(/^[0-9a-zA-Z]+$/)) {
```
Comment (djones) [inline src/environment.js]: According to Git's [docs on Git Objects](https://git-scm.com/book/id/v2/Git-Internals-Git-Objects), a commit sha is "40-character checksum hash". So probably should max out at that size.

In addition, you could enforce the length into the regex here. Something like this:

```
if (!commitSha.match(/^[0-9a-zA-Z]{0,40}$/)) {
```

Comment (timhaines) [reply src/environment.js]: Yeah, I was aiming to provide a little leeway, but not a lot.  Sometimes the "commitSha" is `HEAD` too.

File: src/environment.js [hunk 2]
```diff
@@ -26,6 +37,66 @@ class Environment {
...
+      return '';
+    }
+
+    const child_process = require('child_process');
+    const format = GIT_FORMAT_LINES.join('%n'); // git show format uses %n for newlines.
+    const cmd = `git show ${commitSha} --quiet --format="${format}"`;
```
Comment (djones) [inline src/environment.js]: Nice to have this so locked down. Commit sha goes through regex and format is a const. :+1:

File: src/environment.js [hunk 3]
```diff
@@ -53,34 +124,66 @@ class Environment {
...
+    if (result == '') {
+      // Branch not specified
+      result = null;
+    }
+
+    return result;
+  }
+
+  rawBranch() {
+    const child_process = require('child_process');
```
Comment (djones) [inline src/environment.js]: I notice this same sort of setup twice. Optionally consider creating a method like `execGit`.

Something like this:

```javascript
execGit(command) {
  try {
    return child_process
      .execSync(`git ${cmd}`)
      .toString()
      .trim();
  } catch (error) {
    return '';
  }
}
```

Then you could have

```javascript
execGit('rev-parse --abbrev-ref HEAD')
```

Comment (timhaines) [reply src/environment.js]: Good call!  Will do.

**Pull request merged**